src/fold/mod.rs

```diff
@@ -136,7 +157,32 @@ impl<T: Fold> Fold for Option<T> {
 impl Fold for Ty {
     type Result = Self;
     fn fold_with(&self, folder: &mut Folder, binders: usize) -> Result<Self::Result> {
-        folder.fold_ty(self, binders)
+        match *self {
+            Ty::Var(depth) => if depth >= binders {
+                folder.fold_free_existential_ty(depth - binders, binders)
+            } else {
+                Ok(Ty::Var(depth))
+            },
+            Ty::Apply(ref apply) => {
```

Comment:
I think it was this match-arm that made me stumble. Your solution looks certainly better than everything I came up with :)